### DISCUSSION TITLE
The LSP should generate diagnostics from the formatter

### DISCUSSION THREAD

**Discussion by In3luki** [biomejs/biome#809] [Ideas]
For feature parity with the Prettier vscode extension it would be nice if Biome could also highlight lines that need formatting:

![prettier](https://github.com/biomejs/biome/assets/41452412/9649110d-1603-4182-aa7a-8ea944a5eee3)

Edit: Looking at the screenshot again it may actually be the ESLint extension that emits the diagnostics but my point still stands. 😅

**Comment by ematipico** [comment]
Mmmmh, are you sure that's what you want? Apart from the feature parity, that diagnostic isn't that great because the text is cut. 

We usually tend to use LSP diagnostics for messages that are more actionable. Do you have any other suggestions to make the diagnostic better?

**Comment by In3luki** [comment]
I don't think the text is cut off?  The dots represent the indenation (as spaces) that will be added by the formatter. I might misunderstand what you mean.
For me personally it's a convenient way to see what is wrong with a file before it is altered by e.g. auto formatting on save.

If I want to ignore a problem I can add a `// eslint-disable-next-line prettier/prettier` comment via the `Quick Fix` context menu without having to undo the formatting of the whole file after it was formatted.
These are all the  `Quick Fix` options for reference:
![image](https://github.com/biomejs/biome/assets/41452412/0a23e5d8-f829-492f-9bf4-c00fa2f4e22a)

**Comment by martinbiard** [comment]
I was also looking for this for the Jetbrains IDEs as I prefer to disable any automatic formatting and see the issues flagged in the editor instead. The visual flagging of code that needs a reformat lets me know that there are issues and I remain in control to decide when to reformat the whole file, or only part of the code, using the Ctrl-Alt-L shortcut. When you are used to this visual feedback, it is hard to consider migrating to biome, which doesn't provide it.

Below is how I like to see those formatting issues in my editor.

![eslint_prettier](https://github.com/biomejs/biome/assets/5204353/497fc395-e453-4649-85ad-f4cad7df1e1f)

**Comment by mdrobny** [comment]
Me and people at my company are also surprised to see that there is no visual information (highlight) about lines that need formatting.
In my opinion it's useful to see visually what code is following the formatting rules and what code violates it. 

I am not used to switching between editor and terminal to check if code is formatted correctly. Maybe I need to learn it or enable formatting on file save but it's not obvious approach to me currently.
It would be also important when introducing new developers to the codebase so they can quickly learn about the formatting rules.

Biome is awesome 🚀  and I hope we can start using it in all projects but such small things are sometimes big blockers 😭

**Reply by ematipico** [reply]
The Biome formatter is closer to how editors (hence LSP) use it. Users configure the formatting as they wish (on save, on type, on select, etc.), and the editor takes care of the rest. The CLI is a safety net if users want to enforce it in CI.

We think formatting is not meant to be used as a lint rule.

Plus, having formatting as a lint rule could cause the following "problems": 
- Degraded DX where multiple lines *might* have a red squiggle because something isn't formatted
- Having multiple actions that do the same thing could degrade the performance of the LSP. Multiple actions: the LSP formatting action and the *inline* formatting action
- Formatter and Linting suppression systems have different semantics. If we mix them, it could cause more confusion

Also, we try to look at other ecosystems (not just JS) for standards, and it seems that this is the way to go for Biome. 

I'm sorry if this is a massive blocker for you; I just hope that you're open to embracing a different vision of formatting and linting.

**Reply by dionysiusmarquis** [reply]
I‘m not very deep into the lsp topic, but I‘m also visualising prettier formatting via eslint lsp. Eslint marks the error/warn/hint (based on level setting) as "prettier/…". Therefore It‘s obvious that this is formatter related.

An option to optionally visualise biome messages in set level alongside linting would be nice.

**Reply by Kolobamanacas** [reply]
@ematipico I'm sorry to jump in after a couple of years, but may I ask: is this decision final and will no changes be made to address this feature request? I finally tried Biome as it now seems mature enough, but lack of formatting issues highlighting seems like a blocker for my team too.

- If by DX you mean developer experience, I don't understand how exactly highlighting formatting issues degrades it. One of developer's goals is to deliver code which is aligned with project's rule set, including styling. If it is not aligned without highlighting the developer would never know. Auto-formatting is not an savior for the having-control reasons mentioned by @martinbiard. Post-formatting is even in that sense as the code left by the developer differs from the code which ends up in a code base, which leads to confusion.
- I'm also not very strong with linter/formatter internals, but it is quite hard to grasp the argument about multiple actions for the same thing. Why would they be the same? Currently there is literally no action for stylistic mismatches. From the external point of view if within one file there are multiple issues, for example, a value is only used as a type and imports are in the wrong order, this means that Biome checks the file against the list of pre-defined rules and each has its own individual action to apply a fix and all of them can be fixed with "fix all". What difference does it make if there will be more rules like formatting-related ones?
- I could be wrong, but I thought that the whole idea of Biome was to (finally) rule them all and provide one unified tool to do both linting and formatting. And it turns out it does only the half job for the formatting. I'm not in place to give advice on this one, but I believe there must be some way to handle this elegantly. As a thought, to unify semantics for formatter and linter suppression systems.

I would personally add that visualizing code mismatches is in a way more important for some teams that the actual formatting. Of course without the latter it doesn't make sense, but only two of them together provide the most comfortable experience. And for the other teams who prefer the way it is now, it could always be made configurable.

I really hope that this decision is not yet set in stone and something could be changed. I badly wish to switch to something better than ESLint/Prettier.

**Reply by ematipico** [reply]
The decision isn't final.

The degradation is mostly based on what I've seen with the prettier extension and what it's need from Biome to deliver the feature. So it's my impression, not something that is certain. 

Regardless, this request has still 7 upvotes, and we have feature requests that have more, so we need to give them more priority.
